src/platform/web/ui/general/LazyListView.js

```diff
@@ -0,0 +1,250 @@
+import {el} from "./html.js";
+import {mountView} from "./utils.js";
+import {insertAt, ListView} from "./ListView.js";
+
+class ItemRange {
+    constructor(topCount, renderCount, bottomCount) {
+        this.topCount = topCount;
+        this.renderCount = renderCount;
+        this.bottomCount = bottomCount;
+    }
+
+    contains(range) {
+        // don't contain empty ranges
+        // as it will prevent clearing the list
+        // once it is scrolled far enough out of view
+        if (!range.renderCount && this.renderCount) {
+            return false;
+        }
+        return range.topCount >= this.topCount &&
+            (range.topCount + range.renderCount) <= (this.topCount + this.renderCount);
+    }
+
+    containsIndex(idx) {
+        return idx >= this.topCount && idx <= (this.topCount + this.renderCount);
+    }
+
+    expand(amount) {
+        // don't expand ranges that won't render anything
+        if (this.renderCount === 0) {
+            return this;
+        }
+
+        const topGrow = Math.min(amount, this.topCount);
+        const bottomGrow = Math.min(amount, this.bottomCount);
+        return new ItemRange(
+            this.topCount - topGrow,
+            this.renderCount + topGrow + bottomGrow,
+            this.bottomCount - bottomGrow,
+        );
+    }
+
+    totalSize() {
+        return this.topCount + this.renderCount + this.bottomCount;
+    }
+
+    normalize(idx) {
+        /*
+        map index from list to index in rendered range
+        eg: if the index range of this._list is [0, 200] and we have rendered
+        elements in range [50, 60] then index 50 in list must map to index 0
+        in DOM tree/childInstance array.
+        */
+        return idx - this.topCount;
+    }
+}
+
+export class LazyListView extends ListView {
+    constructor({itemHeight, overflowMargin = 5, overflowItems = 20,...options}, childCreator) {
+        super(options, childCreator);
+        this._itemHeight = itemHeight;
+        this._overflowMargin = overflowMargin;
+        this._overflowItems = overflowItems;
+    }
+
+    _getVisibleRange() {
+        const length = this._list ? this._list.length : 0;
+        const scrollTop = this._parent.scrollTop;
+        const topCount = Math.min(Math.max(0, Math.floor(scrollTop / this._itemHeight)), length);
+        const itemsAfterTop = length - topCount;
+        const visibleItems = this._height !== 0 ? Math.ceil(this._height / this._itemHeight) : 0;
+        const renderCount = Math.min(visibleItems, itemsAfterTop);
+        const bottomCount = itemsAfterTop - renderCount;
+        return new ItemRange(topCount, renderCount, bottomCount);
+    }
+
+    _renderIfNeeded() {
+        const range = this._getVisibleRange();
+        const intersectRange = range.expand(this._overflowMargin);
+        const renderRange = range.expand(this._overflowItems);
+        const listHasChangedSize = !!this._renderRange && this._list.length !== this._renderRange.totalSize();
```

Comment:
`listHasChangedSize` can be removed here and in the check below, as any changes to the list will be handled in the onAdd/onMove/onRemove/... handlers.

Reply:
`listHasChangedSize` is removed but we still need an option to force render for the time being until we optimize onAdd/onRemove.